### PR TITLE
Wrap the beego router and provide a unified view for users to register routes

### PR DESCRIPTION
**Pull request by ywk253100**
Wrap the beego router and provide a unified view for users to register routes

Signed-off-by: Wenkai Yin <yinw@vmware.com>

### CODE REVIEW

File: src/core/main.go [hunk 1]
```diff
@@ -251,7 +252,7 @@ func main() {
 	beego.InsertFilter("/*", beego.BeforeRouter, filter.SecurityFilter)
 	beego.InsertFilter("/*", beego.BeforeRouter, filter.ReadonlyFilter)
```
Comment (steven-zou) [inline src/core/main.go]: TODO REMOVE?

File: src/server/router/router.go [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+			beego.Delete(r.path, filterFunc)
+		case http.MethodOptions:
+			beego.Options(r.path, filterFunc)
+		}
+	}
+}
+
+// HandlerFunc sets the handler function that handles the request
+func (r *Route) HandlerFunc(f http.HandlerFunc) {
+	r.Handler(http.HandlerFunc(f))
```
Comment (steven-zou) [inline src/server/router/router.go]: `f` has type `http.HandlerFunc`, why need a conversion again?

File: src/server/router/router.go [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+// Route stores the information that matches a request
+type Route struct {
+	methods     []string
+	path        string
+	middlewares []middleware.Middleware
+}
+
+// Methods sets the methods that the route matches
+func (r *Route) Methods(methods ...string) *Route {
```
Comment (steven-zou) [inline src/server/router/router.go]: As you have a method chain way, why need to pass method array here? 

`Method(m string)` seems is enough.

Comment (steven-zou) [reply src/server/router/router.go]: `WithMethod` is more suitable as it is op.

File: src/server/router/router.go [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+	r.methods = methods
+	return r
+}
+
+// Path sets the path that the route matches. Path uses the beego router path pattern
+func (r *Route) Path(path string) *Route {
```
Comment (steven-zou) [inline src/server/router/router.go]: `WithPath()`?

File: src/server/router/router.go [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+	r.path = path
+	return r
+}
+
+// Middleware sets the middleware that executed when handling the request
+func (r *Route) Middleware(middleware middleware.Middleware) *Route {
```
Comment (steven-zou) [inline src/server/router/router.go]: `WithMiddleware()`?

I think you just need to keep a front pointer` of the middleware, here 

`m(middleware)` is ok for chaining them together.

File: src/server/router/router.go [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+	r.middlewares = append(r.middlewares, middleware)
+	return r
+}
+
+// Handler sets the handler that handles the request
+func (r *Route) Handler(handler http.Handler) {
+	filterFunc := beego.FilterFunc(func(ctx *beegocontext.Context) {
+		ctx.Request = ctx.Request.WithContext(
+			context.WithValue(ctx.Request.Context(), contextKeyInput{}, ctx.Input))
+		// TODO remove the WithMiddlewares?
```
Comment (steven-zou) [inline src/server/router/router.go]: If middleware has been chained together, then code here is not required.

File: src/server/router/router.go [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+	}
+}
+
+// HandlerFunc sets the handler function that handles the request
+func (r *Route) HandlerFunc(f http.HandlerFunc) {
+	r.Handler(http.HandlerFunc(f))
+}
+
+// GetInput returns the input object from the context
+func GetInput(context context.Context) (*beegocontext.BeegoInput, error) {
```
Comment (steven-zou) [inline src/server/router/router.go]: Is it possible to provide more wrapped easy-using methods like Parameters(k string), not just a high-level entry object?

File: src/server/router/router.go [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+	}
+	input, ok := context.Value(contextKeyInput{}).(*beegocontext.BeegoInput)
+	if !ok {
+		return nil, errors.New("input not found in the context")
+	}
+	return input, nil
+}
+
+// Middleware registers the global middleware that executed for all requests that match the path
+func Middleware(path string, middleware middleware.Middleware) {
```
Comment (steven-zou) [inline src/server/router/router.go]: Is it possible to provide a route with path `*` to cover this requirement with the same Route model and flow?

File: src/server/router/router.go (hunk 7 above)
Comment (ywk253100) [reply src/server/router/router.go]: `Param` function added

File: src/server/router/router.go (hunk 8 above)
Comment (ywk253100) [reply src/server/router/router.go]: Beego doesn't support specify the path, I'll refactor this function after weiwei'PR merged

File: src/core/main.go (hunk 1 above)
Comment (ywk253100) [reply src/core/main.go]: Issue https://github.com/goharbor/harbor/issues/10532 opened to track the work

**Pull request merged**